### DISCUSSION TITLE
"found associated type" when attempting query over a generic component type

### DISCUSSION THREAD

**Discussion by bbarker** [bevyengine/bevy#10387] [Q&A] [answered]
I've been struggling a bit with this and figured I may be going about things the wrong way. I'm trying to write some serialize helper functions and macros that will serialize the component along with the entity given a specific component type, as well as a marker component indicating whether the entity should be serialized.

I've deleted most of the extraneous code:

```rust
pub trait SerializeComponents<'a, E, C, CQ, M>
where
    M: Component,
    C: Component + WorldQuery + Serialize + Clone,
{
    fn serialize<S>(
        &self,
        query: Query<(Entity, C), With<M>>,
        serializer: S,
    ) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut serseq = serializer.serialize_seq(Some(count))?;

        query.iter().map(|(ent, comp)| {
            let ent_comp_tuple: (Entity, C) = (ent, comp);
        });

        serseq.end()
    }
}
```

This results in the error:

```
18 | pub trait SerializeComponents<'a, E, C, CQ, M>
   |                                      - this type parameter
...
61 |             let ent_comp_tuple: (Entity, C) = (ent, comp);
   |                                                     ^^^^ expected type parameter `C`, found associated type
   |
   = note: expected type parameter `C`
             found associated type `<<C as WorldQuery>::ReadOnly as WorldQuery>::Item<'_>`
   = note: you might be missing a type parameter or trait bound

```

**Comment by SkiFire13** [comment, marked as answer]
You likely want to remove the `C: WorldQuery` bound and use `Query<(Entity, &C), With<M>>` instead of your current query.

Explanation of what happened: queries over some `WorldQuery` `C` will yield values of type `<C as WorldQuery>::Item<'w>`. For example if `C` was `&mut T` (for some `T: Component`) then the values yielded will be of type `bevy::ecs::change_detection::Mut<'w, T>`. For `&C` however the query item is also `&C`, so there shouldn't be problems. Also, generally a `Component` `C` is not a valid `WorldQuery`, but `&C` and `&mut C` are, hence why I suggested you to remove the `C: WorldQuery` bound.

**Reply by bbarker** [reply]
Thank you for pointing out that beginner mistake - I'm still getting used to queries and forgot to use a reference component type, and went down a bad path!
